### ISSUE TITLE
Find latest db version

### ISSUE THREAD

**Issue by vigyanhoon** [dfahlander/Dexie.js#277]
Hi,

Here is a scenario.
1) I create a two tables tableA and tableB.
2) I fill some data in both the tables.
3) I delete tableA as its no longer needed by version upgrade.
4) I refresh the browser.
5) Now how can I know which version the DB is at as I intend to open the latest and greatest version of it.
6) I also want to create a new table on the latest version.
7) Basically I want version to be done with dynamic number rather than providing static 1, 2, 3 ... etc.

Thank you.

**Comment by dfahlander** [comment]
I think you need to think differently when working with dexie or indexedDB in general.

Think like this : 

1. How do I want the schema to be? 
2. What is installed on my users browsers currently? (what versions have been published earlier and what stores and indexes)
3. How do I convert existing data to the current format?
4. Are there any unused stores that could be deleted? 

### To answer question 1

Define a new version(prevVer + 1).stores(...  how you want it to be). 

### Question 2

Keep the old version definitions side by side with the latest one to help Dexie make a diff. Never touch them after having published a version publicly. Keep them as they were defined at the time they were latest and greatest version.

### Question 3

If you need to upgrade your data along with the schema changes, append `.upgrade()` to the version.

### Question 4

If an object store is no longer used, specify null for it. If your upgrade function needs to read from that deleted store, declare the deleted stores in yet a new version.

### Summary 

Basically you don't need to know what version is currently installed. Each version you declare reflects a current or previous version. Dexie finds out the version installed and makes appropriate changes.

**Comment by dfahlander** [comment]
That said, it is also possible to open the database dynamically and then find current version using db.verno. See the api reference and samples that covers this (Sample 'open existing database')

**Comment by vigyanhoon** [comment]
When I try to create store after refreshing page I get 'The requested version (10) is less than the existing version (20).' as during updation version was changed to 2. So I need to be able to find out that current version of database is at 2 after page refresh.

-----------------------------------------------------------------------
### I have created a db with some tables:
```
var db = new Dexie('Database');
db.version(1).stores(definitions);
```

### After some time I delete one of the tables like:
```
db.close();
db.version(db.verno+1).stores({
  table1: null
})
```
Db version got upgraded to 2 here.

### Now I refresh page and want to add a table
```
var db = new Dexie('Database'); (db.verno is 0 here)
db.version(_**I dont know latest version here**_).stores(definitions); ??
```

I hope I am making myself clear.

**Comment by dfahlander** [comment]
verno isnt known until you've opened the database in dynamic mode (as the sample 'open existing db' shows).

**Comment by dfahlander** [comment]
https://github.com/dfahlander/Dexie.js/blob/master/samples/open-existing-db/dump-databases.html

**Comment by vigyanhoon** [comment]
Yeah, I got the version after opening DB. Thank you!

However, I am not able to delete a table after I refresh the browser. Can you guide me, please?

If I create a new table and drop it by setting it to null in db.version(db.verno+1), it works beautifully.

However if I create a table and then refresh the browser, and then try to delete the table the same way. It throws -> 'Dexie specification of currently installed DB version is missing'. 

I debugged and found that oldVersionStruct variable was null after refreshing the browser -> that was because versions array didn't have previous version 1.

Does this mean I can not delete a table directly after a page refresh? I don't really want to write first version table schema again after refresh, just want to delete previously created table.

**Comment by dfahlander** [comment]
I still think you might be missing something fundamental with indexedDB,  and that is: You can't change schema without closing and reopening the database in a incremented version number. This applies to any schema change, such as table creation as well as table deletion. Also, you can never downgrade.

In your case it seems you want to create a temporary table for deletion later on. This implies having to close/increment version/reopen each time you want to either drop or create your table.

**Comment by dfahlander** [comment]
Note: frequently closing and reopening the database might result in database being blocked in some situations when there are several browser windows or tabs open in MSIE or edge browser. Do you really need to do it this way? Couldn't you always have that table present and just to table.clear() instead of deleting it?

**Comment by vigyanhoon** [comment]
I think I followed up on closing->incrementing->opening the DB, as it works the first time without refreshing. But  the previous version is lost on browser refresh.

I think I will have to live with clear for now.

**Comment by dfahlander** [comment]
Yes, you cannot downgrade so previous version is always lost by design, as I tried to explain. I'm glad you user clear instead which is a preferred solution since you don't need to upgrade schema frequently.

**Comment by ddb1494** [comment]
const request = indexedDB.open(DB_NAME);
    request.onsuccess = function (event: any) {
      const db = event.target.result;
      const version = db.version;
      console.log('Current database version:', version);
    };

**Comment by ddb1494** [comment]
- Should fix verno always returning 0 initially.

**Comment by dfahlander** [comment]
> const request = indexedDB.open(DB_NAME); request.onsuccess = function (event: any) { const db = event.target.result; const version = db.version; console.log('Current database version:', version); };
> * Should fix verno always returning 0 initially.

That snippet has to wait for the asynchronous request to complete, just the same way `await db.open()` is required before db.verno is available. Finding out the current verno before defining the schema is also achieved using:

```js
const verno = await new Dexie(DB_NAME).open().then((db) => {
  db.close();
  return db.verno;
});
```
